client/modules/components/doc-gen.js

```diff
@@ -0,0 +1,171 @@
+    .replaceAll('"global.', '"')
+    .replaceAll(/\{@link\s+(\S+?)\s*\}/g, "`$1`")
+    .replaceAll(/_\d+"/g, '"')
+
+  await fse.outputFile("dist/docs.json", outfile)
+
+  spinner.succeed("Components documentation JSON generated.")
+}
+
+// credit to:
+// https://github.com/material-svelte/material-svelte/blob/main/tools/svelte-type-generator/src/index.ts
+// for figuring out how to do this.
+// I use the exact same method here, but with different post-processing
+
+async function createProgram(files, map) {
+  const host = ts.createCompilerHost(TS_CONFIG)
+
+  const fileExists = host.fileExists
+  host.fileExists = filename => {
+    if (filename.endsWith(".svelte.tsx")) return true
+    return fileExists(filename)
+  }
+
+  const readFile = host.readFile
+  host.readFile = filename => {
+    if (filename.endsWith(".svelte.tsx")) {
+      return map.get(filename)
+    }
+    return readFile(filename)
+  }
+
+  const writeFile = host.writeFile
+  host.writeFile = (filename, data, BOM, onErr, sourceFiles) => {
+    if (filename.endsWith(".svelte.d.ts")) {
+      // this is a hack, but I couldn't get anything else to cooperate
+      // this makes it so that the props, events, and slots
+      // properties are not optional, which screws up types otherwise
+      data = data
+        .replaceAll("    props?: {", "    props: {")
+        .replaceAll("    events?: ", "    events: ")
+        .replaceAll("    slots?: ", "    slots: ")
+        .replaceAll("    } | undefined;", "    };")
+        .replaceAll(" {} | undefined;", " {};")
+    }
+    return writeFile(filename, data, BOM, onErr, sourceFiles)
+  }
+
+  const tsxFiles = files.map(file => `${file}.tsx`)
+  const shims = "../../node_modules/svelte2tsx/svelte-shims.d.ts"
+  const types = await globby("../misc/types/*.d.ts", { absolute: true })
+
+  const program = ts.createProgram([...tsxFiles, shims, ...types], TS_CONFIG, host)
+
+  return { program, tsxFiles }
+}
+
+// prettier-ignore
+const COMPONENT_REGEX =
+ /export default class (.+)__SvelteComponent_ extends createSvelte2TsxComponent\((.+)\)\s*\{\s*\}/
+
+async function generateTSX(file) {
+  const src = await fs.readFile(file, "utf-8")
+  const tsx = svelte2tsx(src, { filename: file, strictMode: true, isTsFile: true })
+
+  // postprocessing
+  const match = COMPONENT_REGEX.exec(tsx.code)
+  if (match) {
+    const [oldExport, name, props] = match
+
+    const newExport =
+      `export declare class ${name} {` +
+      `\n  constructor(options: IComponentOptions)` +
+      `\n  props: typeof props.props;` +
+      `\n  events: typeof props.events;` +
+      `\n  slots: typeof props.slots;` +
+      `\n}`
+
+    // prettier-ignore
+    const out =
+      `\nexport interface IComponentOptions { target: Element; anchor?: Element; props?: Record<string, any>; context?: Map<any, any>; hydrate?: boolean; intro?: boolean; };` +
+      `\n${tsx.code.replace(oldExport, newExport)};` +
+      `\nconst props = ${props}();`
```

Comment:
Would it make sense / be possible to have this kind of wrapper code in a template file or similar?

Reply:
Not sure, honestly. I don't know what solutions for that kinda thing exist. This code is about as simple as I could get it, lol.